### DISCUSSION TITLE
Is WDI5 safe to use in SAP context

### DISCUSSION THREAD

**Discussion by Vikrant87-lego** [ui5-community/wdi5#549] [General]
Hi All,

I am QA and I tried to use WDI5 tool to automate the Regression flows in our SAP Neo application and i was able to do some of the Flows, which i tried to share it with my ABAP developers and there were few concerning questions asked to me as below:

1. Since WDI5 is an external open source library is it risk to have the dependencies in the project repository?(similar to log4j vulnerabilities)
2. Can WDI5 find elements which have dynamic UI5 properties?(e.g. what if the id of the control changes each time the app is deployed)

also they suggested me to try only SAP approved frameworks like Qunit or OPA5 so are the above concerns valid or its just the developers overthinking?

Regards,
Vikrant

**Comment by vobu** [comment]
Hi Vikrant, `wdi5` is the official e2e test tool for UI5, so mentioned in the docs (e.g. the test scope pyramid in https://ui5.sap.com/#/topic/ab134ef3932c4b42898c79c10341e8b5), and endorsed by the UI5 core team.

All I can further say in this regard is that the cooperation between SAP/UI5 core and the `wdi5` maintainers and its [steering committee](https://github.com/ui5-community/wdi5#steering-committee) is going exceptionally well, with a high degree of mutual trust. It could serve as a lighthouse for future projects where both SAP and its community can benefit from fast feature delivery and quicker innovation cycles.

To address your points:
1) where ever there are dependencies, there's a need to maintain them. This is for the largest part what denotes "technical debt" in the fast-paced Node.js universe. This has nothing to do with SAP or any external libraries, it is just a given aspect of the software world.

2) yes. There's plenty of locators that don't rely on static IDs of a control: https://ui5-community.github.io/wdi5/#/locators

hth, V.

**Comment by Vikrant87-lego** [comment]
Hi Volker,

Thank you very much for the reponse.

I have been facing issue with selectors which are not uniquely identified and the ones which are identified uniquely than the UI5 equivalent methods are not working on them. May be the way i structure the Selectors could be a problem , so is there any tutorial or guide how to work with selectors and specfic UI5 controls like TreeTable etc?

And the Question about the dependencies, in my project repository i have tried to put the WDI5 dependencies in the root project and kept the Spec files under root/webapp/test directory but still the developers have a concerns that it could impact the applications deployment if the tests/test resources reside inside the project directory. Is it a valid concern?

Also i am new to SAP but i have been questioned that if SAP changes anything in future for SAPUI5, version upgrade or anything else, will the test still work in that scenario. will the WDI5 test work in cloud foundary as well?

One last question , is there a dedicated support available if we face issues with WDI5.

Regards,
Vikrant

**Comment by vobu** [comment]
> I have been facing issue with selectors which are not uniquely identified and the ones which are identified uniquely than the UI5 equivalent methods are not working on them. May be the way i structure the Selectors could be a problem , so is there any tutorial or guide how to work with selectors and specfic UI5 controls like TreeTable etc?

there's quite some documentation for locators https://ui5-community.github.io/wdi5/#/locators .
recipes for common use cases are at https://ui5-community.github.io/wdi5/#/recipes

> And the Question about the dependencies, in my project repository i have tried to put the WDI5 dependencies in the root project and kept the Spec files under root/webapp/test directory but still the developers have a concerns that it could impact the applications deployment if the tests/test resources reside inside the project directory. Is it a valid concern?

IMO not really. UI5 apps default test dir location is `/webapp/test/**/*` - see for example https://github.com/SAP-samples/ui5-typescript-helloworld/tree/testing 

> Also i am new to SAP but i have been questioned that if SAP changes anything in future for SAPUI5, version upgrade or anything else, will the test still work in that scenario. will the WDI5 test work in cloud foundary as well?

as we're reusing UI5's Test API "`RecordReplay`", I'll go out on a limb here that only when UI5 itself is broken, then `wdi5` will be broken.
In regards to your question "does `wdi5` work in Cloud Foundry?", this is a bit off unfortunately. `wdi5` itself runs anywhere where there's a Node.js runtime. But I'm assuming you're referring to a _UI5 app_ that runs in Cloud Foundry - and yes, `wdi5` can also access that, even with [built-in authentication](https://ui5-community.github.io/wdi5/#/authentication).

> One last question , is there a dedicated support available if we face issues with WDI5.

yes, there is. Community support works via [the regular GitHub issues](https://github.com/ui5-community/wdi5/issues). 
Dedicated [commercial support is available by these companies](https://github.com/ui5-community/wdi5/blob/main/SUPPORT.md), applied via GitHub Sponsors.

hth, v.
